videoclip.lua

```diff
@@ -431,6 +475,90 @@ function main_menu:update()
+                os_type == 'windows' and 'curl.exe' or 'curl', '-s',
+                '-F', 'reqtype=fileupload',
+                '-F', 'time=' .. config['litterbox_expire'],
+                '-F', 'fileToUpload=@"' .. outfile .. '"',
+                endpoint
+            })
+
+            -- Exit codes in the range [0, 99] are returned by cURL itself.
+            -- Any other exit code means the shell failed to execute cURL.
+            if r.status < 0 or r.status > 99 then
+                notify("Error: Failed to upload. Make sure cURL is installed and in your PATH.", "error", 3)
+                return
+            end
+            if r.status ~= 0 then
+                notify("Error: Failed to upload to " .. config.litterbox and "litterbox.catbox.moe" or "catbox.moe", "error", 2)
+                return
+            end
+
+            -- Copy to clipboard
+            if os_type == 'windows' then
+                local clipboard_command = {
+                    'powershell.exe', '-command',
+                    'Set-Clipboard -Value ' .. r.stdout
+                }
+                
+                local cb = subprocess(clipboard_command)
+                -- local status = mp.command('run ' .. clipboard_command)
+                if cb.status ~= 0 then
+                    notify("Failed to copy URL to clipboard, trying to open in browser instead. (Make sure PowerShell is installed)", "warn", 4)
+                    mp.commandv('run', open_utility, r.stdout)
+                    return
+                end
+            end
+
+            if os_type == 'macos' then
+                local clipboard_command = {
+                    "pbcopy"
+                }
+                local cb = subprocess(clipboard_command, r.stdout)
+                if cb.status ~= 0 then
+                    notify("Failed to copy URL to clipboard, trying to open in browser instead. (Make sure pbcopy is installed)", "warn", 4)
+                    mp.commandv('run', open_utility, r.stdout)
+                    return
+                end
+            end
+
+            if os_type == "linux" then
+                local cb_x11 = string.find(subprocess({
+                    "whereis", "xclip"
+                }).stdout, "/")
+                local cb_wayland = string.find(subprocess({
+                    "whereis", "wl-copy"
+                }).stdout, "/")
+
+                if cb_x11 == nil and cb_wayland == nil then
+                    notify("Failed to copy URL to clipboard, trying to open in browser instead.\n(Make sure xclip or wl-clipboard is installed)", "warn", 6)
+                    mp.commandv('run', open_utility, r.stdout)
+                    return
+                end
+
+                if cb_x11 ~= nil then
+                    mp.command("run /bin/sh -c \"echo " .. r.stdout .. " | xclip -sel clip\"")
+                elseif cb_wayland ~= nil then
+                    mp.command("run wl-copy " .. r.stdout)
+                end
```

Comment:
Just send the text to stdin to avoid problems with quotation marks.
``` lua
local clipboard_command = is_wayland() and "wlcopy" or "xclip -selection clipboard"
local handle = io.popen(clipboard_command, 'w') 
handle:write(text)
handle:close()
```

Reply:
The text shouldn't have any quotes, since it's a URL, but I guess it is technically untrusted so I'll change Linux and MacOS to this. I don't like using `io.popen` for Windows however, as it opens up a command window.

Reply:
Yes, there's no need to use `io.popen` on windows.

Reply:
Another thing about `io.popen` is that there's no easy way to get exit codes without string splicing.

Reply:
I guess calling `which` is still needed to test if wl-copy or xclip are installed.

Reply:
Oh, didn't see you'd already refactored most of it when writing https://github.com/Ajatt-Tools/videoclip/pull/20#discussion_r1241283751, looks awesome.

Yeah I don't think there's a good way to test whether they're installed, and `which` is pretty messy. `io.popen` also returns a valid object even if the command it's running fails, so that's annoying.

Reply:
Oh wow, I did some digging and turns out `io.popen` does in fact give you an exit code upon calling the `handle:close()` method. Specifically, it returns 3 values:
- `success`: `boolean`  
- `exitcode`: `"exit" | "signal"`  
- `code`: `integer`

Seems this'll work after all.

Reply:
6c571a8